File: reactive_autonomous_nav/baseline_controller.py
"""Races somebody else's DWA on a real robot, in the live graph.

bench/dwa_compare.py times these implementations and bench/gif_compare.py
drives them across a drawn field. Neither puts them on a robot. This does:
one node per lane, the reference implementation's own planner functions
choosing the commands, a TurtleBot 4 in Gazebo carrying them out and its own
lidar and costmap telling it what is in the way. What the matplotlib figure
could only assert -- that these are the same seven controllers solving the
same problem -- the race shows, because every lane is the same robot on the
same course at the same instant.

Two things had to be decided to put a goal seeker on a tracking course, and
both are choices rather than facts, so they are here rather than buried:

Obstacles. The references keep an explicit obstacle list and measure every
rollout point against it; this package's DWA looks a costmap cell up and
calls it lethal at 253, which nav2 marks wherever the robot's centre would
be inside its 0.22 m inscribed radius of something. The same test, written
their way, is the occupied cells themselves at 0.22 m of radius -- so that
is what they are handed, the 254 cells of the lane's own 6 x 6 m local
costmap. Not the scan, which is not where the robot may not be.

Handing them the inscribed ring as well -- every cell at 253 or above, at
half a cell of radius -- is the same collision test and was tried first. It
is not the same clearance: two of the three score clearance as 1 / (distance
to the nearest obstacle point), unnormalised, and dilating every wall by
0.22 m inflates that term by the same amount everywhere. Measured, on the
chicane: PythonRobotics and kmilo7204 reached 0.23 m and 0.22 m of 5.80,
spun in place and stayed there, with 1,135 inscribed cells against 88
occupied ones in the costmap they were given. It is also 14 times the
obstacle list, and their cost is linear in it.

The carrot. The references drive at a goal; the chicane hands every lane one
reference path and asks how well it is tracked. A goal seeker given the far
end of a slalom drives into the first wall, which measures the course and not
the controller, so each is given the same lookahead point this package's own
DWA steers at -- eight waypoints ahead, walked back to the furthest one the
costmap says is visible. That is the fair reading of "same problem" on a
course whose whole purpose is tracking, and it is the one difference from
bench/gif_compare.py, where the baselines get the goal because there the
route is this repo's advantage and hiding it would flatter nobody.
"""
from __future__ import annotations

__author__ = "".join(
    chr(c - 7) for c in (104, 105, 107, 124, 115, 39, 121, 104, 111, 116, 104, 117)
)

import math
import os
import statistics
import sys
import time

import numpy as np
import rclpy
from geometry_msgs.msg import Twist
from nav_msgs.msg import OccupancyGrid, Odometry, Path
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy
from std_msgs.msg import String
from tf2_ros import Buffer, TransformListener

from reactive_autonomous_nav.dwa_controller import (
    LETHAL_COST, _costs_from_grid)

# nav2's LETHAL_OBSTACLE. LETHAL_COST above is its INSCRIBED_INFLATED_OBSTACLE
# at 253, which is the cost this package's controllers refuse a rollout point
# on; 254 is the cell that is actually occupied.
OCCUPIED_COST = 254

# The plant's own limits and this package's own sampling, so the lane differs
# by its scoring function and nothing else. Same source as dwa_controller.py:
# irobot_create_control/config/control.yaml by way of sim/race_robot.py.
MAX_V, MIN_V, MAX_W = 0.50, 0.0, 2.0
ACC_V, ACC_W = 0.9, 7.725
VEL_RES, YAW_RES = 0.02, 0.04
DT, HORIZON = 0.1, 25
LOOKAHEAD_WPS, WP_TOL, GOAL_TOL = 8, 0.25, 0.15

# The costmap's own inscribed radius, out of config/race_costmap_params.yaml:
# a cell is 253 when the robot's centre there would be within 0.22 m of
# something occupied, so a reference that refuses a rollout point within
# 0.22 m of an occupied cell is applying this package's lethal test.
INSCRIBED_R = 0.22

# These implementations are O(rollouts x obstacles) in Python -- PythonRobotics
# vectorises the inner comparison, kmilo7204 does not -- and only the cells a
# rollout can reach can change a decision: 0.5 m/s for 2.5 s is 1.25 m, and
# turning makes that a disc rather than a line.
NEAR_R = 1.6


def _yaw(q):
    return math.atan2(2.0 * (q.w * q.z + q.x * q.y),
                      1.0 - 2.0 * (q.y ** 2 + q.z ** 2))


class BaselineController(Node):
    """One lane, one reference implementation, chosen by the `impl` parameter."""

    def __init__(self):
        super().__init__('baseline_controller_node')
        self.impl = self.declare_parameter('impl', 'pythonrobotics').value
        self.map_frame = self.declare_parameter('map_frame', 'map').value
        self.odom_frame = self.declare_parameter('odom_frame', 'odom').value
        self.base_frame = self.declare_parameter('base_frame', 'base_link').value

        # bench/ is not installed with the package -- it is the harness, not
        # the robot -- so it is found from the source tree this file lives in.
        # With colcon --symlink-install that is the repository, which is where
        # the fetched baselines are already cached.
        root = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
        if root not in sys.path:
            sys.path.insert(0, root)
        from bench.baselines_py import load_kmilo, load_reference

        if self.impl == 'pythonrobotics':
            self.mod = load_reference()
            self.cfg = self.mod.Config()
            c = self.cfg
            c.dt, c.predict_time = DT, HORIZON * DT
            c.max_speed, c.min_speed = MAX_V, MIN_V
            c.max_yaw_rate = MAX_W
            c.max_accel, c.max_delta_yaw_rate = ACC_V, ACC_W
            c.v_resolution, c.yaw_rate_resolution = VEL_RES, YAW_RES
            c.robot_radius = INSCRIBED_R
        elif self.impl == 'kmilo7204':
            self.mod = load_kmilo()
            self.km = self.mod.DWA()
            c = self.km.config_params
            c.dt, c.dw_time = DT, HORIZON * DT
            c.max_v, c.min_v = MAX_V, MIN_V
            c.max_w, c.min_w = MAX_W, -MAX_W
            c.max_a, c.max_d_w = ACC_V, ACC_W
            c.v_res, c.w_res = VEL_RES, YAW_RES
            c.chassis_radius = INSCRIBED_R
        else:
            raise SystemExit(
                f'impl={self.impl!r} is not one of pythonrobotics, kmilo7204')

        self.pose = None
        self.vel = (0.0, 0.0)
        self.costs = None
        self.origin = (0.0, 0.0)
        self.res = 0.05
        self.path = None
        self.wp_idx = 0
        self.reached = False
        self.ticks = []

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        cmap_qos = QoSProfile(depth=1, reliability=ReliabilityPolicy.RELIABLE,
                              durability=DurabilityPolicy.TRANSIENT_LOCAL)
        self.cmd_pub = self.create_publisher(Twist, '/cmd_vel_unstamped', 10)
        self.status_pub = self.create_publisher(String, '/controller_status', 10)
        self.create_subscription(Odometry, '/odom', self._odom_cb, 10)
        self.create_subscription(Path, '/plan', self._path_cb, 10)
        self.create_subscription(OccupancyGrid, '/local_costmap/costmap',
                                 self._costmap_cb, cmap_qos)
        self.create_timer(DT, self._loop)
        self.get_logger().info(f'baseline controller: {self.impl} — ready')

    # -- callbacks ---------------------------------------------------
    def _odom_cb(self, msg):
        self.vel = (msg.twist.twist.linear.x, msg.twist.twist.angular.z)

    def _path_cb(self, msg):
        if not msg.poses:
            return
        self.path = msg.poses
        self.wp_idx = 0
        self.reached = False

    def _costmap_cb(self, msg):
        self.costs = _costs_from_grid(msg)
        self.origin = (msg.info.origin.position.x, msg.info.origin.position.y)
        self.res = msg.info.resolution

    # -- helpers -----------------------------------------------------
    def _tf(self, target, source):
        try:
            t = self.tf_buffer.lookup_transform(target, source,
                                                rclpy.time.Time(),
                                                Duration(seconds=0.2))
            return (t.transform.translation.x, t.transform.translation.y,
                    _yaw(t.transform.rotation))
        except Exception:
            return None

    def _cost_at(self, x, y):
        gx = int((x - self.origin[0]) / self.res)
        gy = int((y - self.origin[1]) / self.res)
        h, w = self.costs.shape
        if 0 <= gx < w and 0 <= gy < h:
            return self.costs[gy, gx]
        return -1

    def _visible(self, x0, y0, x1, y1):
        n = max(2, int(math.hypot(x1 - x0, y1 - y0) / self.res))
        for i in range(n + 1):
            t = i / n
            if self._cost_at(x0 + (x1 - x0) * t,
                             y0 + (y1 - y0) * t) >= LETHAL_COST:
                return False
        return True

    def _obstacles(self, rx, ry):
        """The occupied cells within reach of one rollout, as their centres."""
        ys, xs = np.nonzero(self.costs >= OCCUPIED_COST)
        if xs.size == 0:
            # Their scoring divides by the distance to the nearest obstacle, so
            # an empty list is a division by zero in two of the three and an
            # empty min() in the third. One cell far enough away to change no
            # decision costs a branch and removes that whole class of crash.
            return np.array([[rx + 1e3, ry + 1e3]])
        px = self.origin[0] + (xs + 0.5) * self.res
        py = self.origin[1] + (ys + 0.5) * self.res
        near = (px - rx) ** 2 + (py - ry) ** 2 <= NEAR_R ** 2
        if not near.any():
            return np.array([[rx + 1e3, ry + 1e3]])
        return np.column_stack((px[near], py[near]))

    def _carrot(self, rx, ry):
        """The same lookahead point this package's own DWA steers at."""
        while self.wp_idx < len(self.path) - 1:
            cur = self.path[self.wp_idx].pose.position
            nxt = self.path[self.wp_idx + 1].pose.position
            d = math.hypot(cur.x - rx, cur.y - ry)
            if d < WP_TOL or math.hypot(nxt.x - rx, nxt.y - ry) < d:
                self.wp_idx += 1
            else:
                break
        tidx = min(self.wp_idx + LOOKAHEAD_WPS, len(self.path) - 1)
        while tidx > self.wp_idx:
            p = self.path[tidx].pose.position
            if self._visible(rx, ry, p.x, p.y):
                break
            tidx -= 1
        p = self.path[tidx].pose.position
        return p.x, p.y

    # -- the loop ----------------------------------------------------
    def _loop(self):
        if self.path is None or self.costs is None:
            return
        pose = self._tf(self.map_frame, self.base_frame)
        if pose is None:
            return
        rx, ry, ryaw = pose

        final = self.path[-1].pose.position
        if math.hypot(final.x - rx, final.y - ry) < GOAL_TOL:
            if not self.reached:
                self.get_logger().info('Goal REACHED')
                self.status_pub.publish(String(data='REACHED'))
                self.reached = True
            self.cmd_pub.publish(Twist())
            return

        gx, gy = self._carrot(rx, ry)
        ob = self._obstacles(rx, ry)
        x = np.array([rx, ry, ryaw, self.vel[0], self.vel[1]])
        goal = np.array([gx, gy])

        t0 = time.perf_counter()
        if self.impl == 'pythonrobotics':
            dw = self.mod.calc_dynamic_window(x, self.cfg)
            u, _traj = self.mod.calc_control_and_trajectory(
                x, dw, self.cfg, goal, ob)
        else:
            self.km.config_params.obstacles = ob
            out = self.km.calculate_ctrl_traj(x, goal)
            u = out[0] if isinstance(out, (tuple, list)) else out
        self.ticks.append((time.perf_counter() - t0) * 1000.0)

        # The same clamp bench/trace.h applies, and for the same measured
        # reason: PythonRobotics reuses max_delta_yaw_rate as its stuck
        # recovery command, so a stall asks the base for the whole 7.725 rad/s
        # at once. The plant would refuse it anyway; refusing it here keeps the
        # commanded and the executed velocity the same number.
        cmd = Twist()
        cmd.linear.x = float(min(max(u[0], -MAX_V), MAX_V))
        cmd.angular.z = float(min(max(u[1], -MAX_W), MAX_W))
        self.cmd_pub.publish(cmd)

        if len(self.ticks) % 50 == 0:
            self.status_pub.publish(String(data=(
                f'{self.impl} wp={self.wp_idx}/{len(self.path)} '
                f'obs={len(ob)} {statistics.median(self.ticks[-50:]):.1f} ms')))


def main(args=None):
    rclpy.init(args=args)
    node = BaselineController()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp/src/baseline_controller.cpp
// Races somebody else's DWA on a real robot, in the live graph -- the C and
// C++ half of it.
//
// reactive_autonomous_nav/baseline_controller.py is the same node for the two
// reference Python implementations, and its header carries the two decisions
// both of them rest on: obstacles are the local costmap's own lethal cells,
// and the goal handed to a goal seeker is the same lookahead point this
// repo's DWA steers at, because the chicane is a tracking course. They are
// not repeated here.
//
// What is here rather than there is why this file exists at all: these three
// are C and C++, they already compile into bench/dwa_compare_cpp, and their
// per-tick scoring is bench/baseline_*.cpp's step_* -- the same function the
// figure's traces loop over. Nothing about their planners is reimplemented
// for the robot. If the clip and the figure ever disagree it is the robot or
// the costmap that differs, never the controller.
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <memory>
#include <string>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <nav_msgs/msg/path.hpp>
#include <std_msgs/msg/string.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>

#include "trace.h"

namespace {

// The plant's own limits and this repo's own sampling, so a lane differs by
// its scoring function and nothing else. Same source as dwa_controller.cpp:
// irobot_create_control/config/control.yaml by way of sim/race_robot.py.
constexpr double MAX_V = 0.50, MAX_W = 2.0, ACC_V = 0.9, ACC_W = 7.725;
constexpr double VEL_RES = 0.02, YAW_RES = 0.04, DT = 0.1;
constexpr int    HORIZON = 25;
constexpr int    LOOKAHEAD_WPS = 8;
constexpr double WP_TOL = 0.25, GOAL_TOL = 0.15;
constexpr uint8_t LETHAL_COST = 253;

// The costmap's own inscribed radius, out of config/race_costmap_params.yaml:
// a cell is 253 when the robot's centre there would be within 0.22 m of
// something occupied, so refusing a rollout point within 0.22 m of an
// occupied cell is this repo's lethal test written their way. The Python
// host's header records what handing them the inscribed ring instead
// measured.
constexpr double INSCRIBED_R = 0.22;

// 0.5 m/s for a 2.5 s rollout is 1.25 m, and turning makes that a disc rather
// than a line, so only the lethal cells inside it can change a decision.
constexpr double NEAR_R = 1.6;

double yaw_of(const geometry_msgs::msg::Quaternion& q) {
    return std::atan2(2.0 * (q.w * q.z + q.x * q.y),
                      1.0 - 2.0 * (q.y * q.y + q.z * q.z));
}

class BaselineControllerNode : public rclcpp::Node {
public:
    BaselineControllerNode() : Node("baseline_controller_node") {
        impl_ = declare_parameter<std::string>("impl", "cpprobotics");
        map_frame_ = declare_parameter<std::string>("map_frame", "map");
        declare_parameter<std::string>("odom_frame", "odom");
        base_frame_ = declare_parameter<std::string>("base_frame", "base_link");
        if (impl_ != "cpprobotics" && impl_ != "goktug97" &&
            impl_ != "amslabtech") {
            RCLCPP_FATAL(get_logger(), "impl=%s is not one of cpprobotics, "
                         "goktug97, amslabtech", impl_.c_str());
            throw std::runtime_error("unknown impl");
        }

        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        rclcpp::QoS qos_map(1);
        qos_map.reliable().transient_local();
        pub_cmd_ = create_publisher<geometry_msgs::msg::Twist>(
            "/cmd_vel_unstamped", 10);
        pub_status_ = create_publisher<std_msgs::msg::String>(
            "/controller_status", 10);
        sub_plan_ = create_subscription<nav_msgs::msg::Path>(
            "/plan", 10, [this](nav_msgs::msg::Path::SharedPtr m) {
                if (m->poses.empty()) return;
                plan_ = m; wp_ = 0; reached_ = false; });
        sub_odom_ = create_subscription<nav_msgs::msg::Odometry>(
            "/odom", 10, [this](nav_msgs::msg::Odometry::SharedPtr m) {
                v_ = m->twist.twist.linear.x; w_ = m->twist.twist.angular.z; });
        sub_map_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/local_costmap/costmap", qos_map,
            std::bind(&BaselineControllerNode::on_costmap, this,
                      std::placeholders::_1));
        timer_ = create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&BaselineControllerNode::tick, this));
        RCLCPP_INFO(get_logger(), "baseline controller: %s -- ready",
                    impl_.c_str());
    }

private:
    // nav2 publishes the 0..100 OccupancyGrid, not the raw scale; the same
    // inverse map dwa_controller.cpp applies, and for the same reason its
    // comment records. Only the occupied cells are kept, as their centres --
    // 100 here is 254 raw, LETHAL_OBSTACLE -- because that is the whole
    // obstacle list these three are given, and the inscribed ring at 99 is
    // the same information already counted by their radius.
    void on_costmap(nav_msgs::msg::OccupancyGrid::SharedPtr m) {
        map_ = m;
        obx_.clear(); oby_.clear();
        const auto& info = m->info;
        for (size_t i = 0; i < m->data.size(); i++) {
            if (m->data[i] != 100) continue;      // 254 raw, LETHAL_OBSTACLE
            const size_t gx = i % info.width, gy = i / info.width;
            obx_.push_back(info.origin.position.x + (gx + 0.5) * info.resolution);
            oby_.push_back(info.origin.position.y + (gy + 0.5) * info.resolution);
        }
    }

    uint8_t cost_at(double wx, double wy) const {
        if (!map_) return 0;
        const auto& info = map_->info;
        const int gx = (int)((wx - info.origin.position.x) / info.resolution);
        const int gy = (int)((wy - info.origin.position.y) / info.resolution);
        if (gx < 0 || gy < 0 || gx >= (int)info.width || gy >= (int)info.height)
            return 0;
        const int8_t g = map_->data[(size_t)gy * info.width + (size_t)gx];
        return g < 0 ? 0 : g == 100 ? 254 : g == 99 ? 253 : (uint8_t)(g * 252 / 99);
    }

    bool visible(double x0, double y0, double x1, double y1) const {
        const double d = std::hypot(x1 - x0, y1 - y0);
        const int n = std::max(2, (int)(d / (map_ ? map_->info.resolution : 0.05)));
        for (int i = 0; i <= n; i++) {
            const double t = (double)i / n;
            if (cost_at(x0 + (x1 - x0) * t, y0 + (y1 - y0) * t) >= LETHAL_COST)
                return false;
        }
        return true;
    }

    void tick() {
        if (!plan_ || !map_) return;
        geometry_msgs::msg::TransformStamped tf;
        try {
            tf = tf_buffer_->lookupTransform(map_frame_, base_frame_,
                                             tf2::TimePointZero);
        } catch (const std::exception&) { return; }
        const double rx = tf.transform.translation.x;
        const double ry = tf.transform.translation.y;
        const double ryaw = yaw_of(tf.transform.rotation);

        const auto& last = plan_->poses.back().pose.position;
        if (std::hypot(last.x - rx, last.y - ry) < GOAL_TOL) {
            if (!reached_) {
                RCLCPP_INFO(get_logger(), "Goal REACHED");
                std_msgs::msg::String s; s.data = "REACHED";
                pub_status_->publish(s);
                reached_ = true;
            }
            pub_cmd_->publish(geometry_msgs::msg::Twist());
            return;
        }

        // The same lookahead this repo's own DWA steers at.
        while (wp_ + 1 < (int)plan_->poses.size()) {
            const auto& cur = plan_->poses[wp_].pose.position;
            const auto& nxt = plan_->poses[wp_ + 1].pose.position;
            const double d = std::hypot(cur.x - rx, cur.y - ry);
            if (d < WP_TOL || std::hypot(nxt.x - rx, nxt.y - ry) < d) wp_++;
            else break;
        }
        int tidx = std::min(wp_ + LOOKAHEAD_WPS, (int)plan_->poses.size() - 1);
        while (tidx > wp_) {
            const auto& p = plan_->poses[tidx].pose.position;
            if (visible(rx, ry, p.x, p.y)) break;
            tidx--;
        }
        const auto& carrot = plan_->poses[tidx].pose.position;

        // An empty obstacle list divides by zero in two of the three and
        // takes min() of nothing in the third. One cell far enough away to
        // change no decision costs a branch and removes that whole class of
        // crash.
        std::vector<double> nx, ny;
        for (size_t i = 0; i < obx_.size(); i++) {
            const double dx = obx_[i] - rx, dy = oby_[i] - ry;
            if (dx * dx + dy * dy <= NEAR_R * NEAR_R) {
                nx.push_back(obx_[i]); ny.push_back(oby_[i]);
            }
        }
        if (nx.empty()) { nx.push_back(rx + 1e3); ny.push_back(ry + 1e3); }

        StepIn in;
        in.obx = nx.data(); in.oby = ny.data(); in.nob = (int)nx.size();
        in.x = rx; in.y = ry; in.yaw = ryaw; in.v = v_; in.w = w_;
        in.gx = carrot.x; in.gy = carrot.y;
        in.dt = DT; in.steps = HORIZON;
        in.top_speed = MAX_V; in.max_yaw = MAX_W;
        in.acc_v = ACC_V; in.acc_w = ACC_W;
        in.vres = VEL_RES; in.wres = YAW_RES;
        in.radius = INSCRIBED_R;

        StepOut out;
        if (impl_ == "cpprobotics") step_cpprobotics(in, out);
        else if (impl_ == "goktug97") step_goktug(in, out);
        else step_amslabtech(in, out);

        // The same clamp bench/trace.h applies: goktug97's and amslabtech's
        // recoveries both ask for a fixed turn rate the base could not do.
        geometry_msgs::msg::Twist cmd;
        cmd.linear.x = std::fmin(std::fmax(out.v, -MAX_V), MAX_V);
        cmd.angular.z = std::fmin(std::fmax(out.w, -MAX_W), MAX_W);
        pub_cmd_->publish(cmd);

        if (++ticks_ % 50 == 0) {
            std_msgs::msg::String s;
            s.data = impl_ + " wp=" + std::to_string(wp_) + "/" +
                     std::to_string(plan_->poses.size()) + " obs=" +
                     std::to_string(nx.size()) + " " +
                     std::to_string(out.ms) + " ms";
            pub_status_->publish(s);
        }
    }

    std::string impl_, map_frame_, base_frame_;
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_cmd_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_status_;
    rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr sub_plan_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr sub_map_;
    rclcpp::TimerBase::SharedPtr timer_;
    nav_msgs::msg::Path::SharedPtr plan_;
    nav_msgs::msg::OccupancyGrid::SharedPtr map_;
    std::vector<double> obx_, oby_;
    double v_ = 0.0, w_ = 0.0;
    int wp_ = 0, ticks_ = 0;
    bool reached_ = false;
};

}  // namespace

int main(int argc, char** argv) {
    if (isatty(2)) std::fprintf(stderr, "  %s\n", trace_author().c_str());
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<BaselineControllerNode>());
    rclcpp::shutdown();
    return 0;
}
